crates/matrix-sdk/src/test_utils/mocks/mod.rs

```diff
@@ -3230,6 +3230,175 @@ impl<'a> MockEndpoint<'a, LoginEndpoint> {
+    ///     let result = client
+    ///         .matrix_auth()
+    ///         .login_username("example", "wordpass")
+    ///         .send()
+    ///         .await
+    ///         .unwrap();
+    ///
+    ///     assert!(
+    ///         result.access_tokesn.unwrap() == "qwerty",
+    ///         "wrong access token in response"
+    ///     );
+    ///     assert!(
+    ///         result.device_id.unwrap() == "DEADBEEF",
+    ///         "wrong device id in response"
+    ///     );
+    ///     assert!(
+    ///         result.user_id.unwrap() == "@cheeky_monkey:matrix.org",
+    ///         "wrong user id in response"
+    ///     );
+    /// }
+    /// ```
+    pub fn ok_with(self, response: LoginResponseTemplate200) -> MatrixMock<'a> {
+        if response.homeserver_url.is_none() && response.identity_url.is_some() {
+            panic!("identity uri given but not homeserver uri. homeserver uri required for well known field.")
+        }
+
+        let identity = if response.identity_url.is_none() {
+            None
+        } else {
+            Some(json!({
+                "base_url": response.identity_url
+            }))
+        };
+
+        let well_known = if response.homeserver_url.is_none() {
+            None
+        } else {
+            Some(json!({
+                "m.homeserver": {
+                    "base_url": response.homeserver_url
+                },
+                "m.identity_server": identity,
+            }))
+        };
+
+        self.respond_with(ResponseTemplate::new(200).set_body_json(json!({
+            "access_token": response.access_token,
+            "device_id": response.device_id,
+            "user_id": response.user_id,
+            "expires_in": response.expires_in,
+            "refresh_token": response.refresh_token,
+            "well_known": well_known,
+        })))
+    }
+
+    /// Ensures that the body of the request is a superset of the provided
+    /// `body` parameter.
+    pub fn body_matches_partial_json(self, body: Value) -> Self {
+        Self { mock: self.mock.and(body_partial_json(body)), ..self }
+    }
+}
+
+/// A response to a [`LoginEndpoint`] query with status code 200.
+#[derive(Default)]
+pub struct LoginResponseTemplate200 {
+    /// Required: An access token for the account. This access token can then be
+    /// used to authorize other requests.
+    access_token: Option<String>,
+
+    /// Required: ID of the logged-in device. Will be the same as the
+    /// corresponding parameter in the request, if one was specified.
+    device_id: Option<String>,
+
+    /// The lifetime of the access token, in milliseconds. Once the access token
+    /// has expired a new access token can be obtained by using the provided
+    /// refresh token. If no refresh token is provided, the client will need
+    /// to re-log in to obtain a new access token. If not given, the client
+    /// can assume that the access token will not expire.
+    expires_in: Option<u128>,
```

Comment:
Didn't we want to use a `Duration` here? Sorry if I forgot to mention this, but time should always use a stronger type in Rust. 

For this case a `Duration` is fine.